front/src/views/Boxes/components/BoxesActionsAndTable.tsx

```diff
@@ -38,54 +54,251 @@ function BoxesActionsAndTable({
+    () =>
+      selectedBoxes.some(
+        (box) => box.values.shipment !== null && box.values.state === BoxState.MarkedForShipment,
+      ),
+    [selectedBoxes],
+  );
+
   // Move Boxes
-  // TODO: replace following refetchQuery with something directly writing to the cache
-  const moveBoxesAction = useMoveBoxes([
-    {
-      query: BOXES_LOCATIONS_TAGS_SHIPMENTS_FOR_BASE_QUERY,
-      variables: {
-        baseId,
-      },
-    },
-  ]);
+  const moveBoxesAction = useMoveBoxes();
 
   const onMoveBoxes = useCallback(
-    (locationId: string) =>
-      moveBoxesAction.moveBoxes(
-        selectedBoxes.map((box) => (box as unknown as BoxRow).labelIdentifier),
+    async (locationId: string) => {
+      const newAlerts: IAlerts[] = [];
+      const movableLabelIdentifiers = selectedBoxes
+        .filter(
+          (box) =>
+            ![BoxState.Receiving, BoxState.MarkedForShipment, BoxState.InTransit].includes(
+              box.values.state,
+            ),
+        )
+        .map((box) => box.values.labelIdentifier);
+
+      const boxCountInShipmentStates = selectedBoxes.length - movableLabelIdentifiers.length;
+      if (boxCountInShipmentStates > 0) {
+        newAlerts.push({
+          id: alerts.length,
+          status: "info",
+          message: `Cannot move ${
+            boxCountInShipmentStates === 1 ? "a box" : `${boxCountInShipmentStates} boxes`
+          } in shipment states.`,
+        });
+      }
+      const moveBoxesResult = await moveBoxesAction.moveBoxes(
+        movableLabelIdentifiers,
         parseInt(locationId, 10),
-      ),
-    [moveBoxesAction, selectedBoxes],
+        true,
+        false,
+      );
+      if (
+        moveBoxesResult.failedLabelIdentifiers &&
+        moveBoxesResult.failedLabelIdentifiers.length > 0
+      ) {
+        newAlerts.push({
+          id: alerts.length,
+          status: "error",
+          message: `Could not move ${
+            moveBoxesResult.failedLabelIdentifiers.length === 1
+              ? "a box"
+              : `${moveBoxesResult.failedLabelIdentifiers.length} boxes`
+          }. Try again?`,
+        });
+      }
+
+      setAlerts([...alerts, ...newAlerts]);
+    },
+    [alerts, moveBoxesAction, selectedBoxes],
+  );
+
+  // Assign to Shipment
+  const { assignBoxesToShipment, isLoading: isAssignBoxesToShipmentLoading } =
+    useAssignBoxesToShipment();
+
+  const onAssignBoxesToShipment = useCallback(
+    async (shipmentId: string) => {
+      const newAlerts: IAlerts[] = [];
+      const assignBoxesToShipmentResult = await assignBoxesToShipment(
+        shipmentId,
+        selectedBoxes.map((box) => box.values as IBoxBasicFields),
+        true,
+        false,
+      );
+      if (
+        assignBoxesToShipmentResult.notInStockBoxes &&
+        assignBoxesToShipmentResult.notInStockBoxes.length > 0
+      ) {
+        newAlerts.push({
+          id: alerts.length,
+          status: "info",
+          message: `Cannot assign ${
```

Comment:
what happens if 10 of those boxes are in stock, and 10 of those are not? Does the instock move succeed? @HaGuesto